### PR TITLE
Edition-level links

### PR DESCRIPTION
**Pull request by thomasleese**
Implementation of edition-level links which solves [the 'draft link' problem][0].

[0]: https://trello.com/c/Yd7RZaaT/870-edition-link-workflow-3

### CODE REVIEW

File: app/models/previously_published_item.rb [hunk 1]
```diff
@@ -57,5 +65,8 @@ def previous_base_path
 
     def routes
     end
+
+    def previously_published_item
```
Comment (dougdroper) [inline app/models/previously_published_item.rb]: ?

File: app/commands/v2/put_content.rb [hunk 2]
```diff
@@ -40,6 +40,19 @@ def update_content_dependencies(edition)
         update_last_edited_at(edition, payload[:last_edited_at])
         ChangeNote.create_from_edition(payload, edition)
         Action.create_put_content_action(edition, document.locale, event)
+        create_links(edition)
+      end
+
+      def create_links(edition)
+        edition.copy_links_from(previously_published_item.links)
```
Comment (kevindew) [inline app/commands/v2/put_content.rb]: I think we should only copy links if `payload[:links]` is null. Treat it as a replacement if it is provided or fallback if not provided.

**Comment by kevindew** [comment]
Ah [this double](https://github.com/alphagov/publishing-api/blob/master/spec/commands/v2/put_content_spec.rb#L5-L7) is why it's not tripping up schema tests - we should probably get rid of that.

**Comment by kevindew** [comment]
Once you rebase this it'll go ❌  as we'll need to update the schemas I'm afraid.

File: app/commands/v2/put_content.rb [hunk 3]
```diff
@@ -44,9 +44,11 @@ def update_content_dependencies(edition)
       end
 
       def create_links(edition)
-        edition.copy_links_from(previously_published_item.links)
+        if payload[:links].nil?
```
Comment (kevindew) [inline app/commands/v2/put_content.rb]: I think it'd be more readable if this had an `if else` to make it clearer that copying and creating are exclusive

**Comment by thomasleese** [comment]
Submitted a PR https://github.com/alphagov/govuk-content-schemas/pull/510 for the schema change.

File: app/presenters/queries/expanded_link_set.rb [hunk 4]
```diff
@@ -3,10 +3,11 @@ module Queries
     class ExpandedLinkSet
       attr_reader :state_fallback_order
 
-      def initialize(content_id:, state_fallback_order:, locale_fallback_order: Edition::DEFAULT_LOCALE)
+      def initialize(content_id:, state_fallback_order:, locale_fallback_order: Edition::DEFAULT_LOCALE, edition_id: nil)
```
Comment (thomasleese) [inline app/presenters/queries/expanded_link_set.rb]: Although it is unnecessary coupling, we believe there would be performance issues with performing the query to find the `edition_id` when it is already available in the places this class is used.

File: app/presenters/queries/expand_dependents.rb [hunk 5]
```diff
@@ -1,9 +1,10 @@
 module Presenters
   module Queries
     class ExpandDependents
-      def initialize(content_id, controller)
+      def initialize(content_id, controller, edition_id)
```
Comment (thomasleese) [inline app/presenters/queries/expand_dependents.rb]: Similar to the comment above.

File: app/models/edition.rb [hunk 6]
```diff
@@ -87,6 +88,14 @@ def base_path_present?
     base_path.present?
   end
 
+  def copy_links_from(other_links)
```
Comment (gpeng) [inline app/models/edition.rb]: Harking back to the original big object avoiding intent of the 'pillars', should this be a concern of `Edition`? As we've got command objects doing things to editions would this be better in there or in its own class? It looks like it's being tested within the `put_content` command tests already.

**Comment by kevindew** [comment]
I haven't had a chance to read in depth so this is a bit of a fleeting comment - but it seems we're now taking locale into account with the edition level links? is this correct / intentional? I'm guessing it pulls in the links of every edition with that content_id.

I feel like since the behaviour of them is somewhat undefined/defined by us we should probably work out some integration tests that define it. What do you think?

**Comment by kevindew** [comment]
closed as it's still a work in progress.

**Pull request closed without merging**